torch/csrc/distributed/rpc/rpc_with_autograd.cpp

```diff
@@ -0,0 +1,161 @@
+#include <torch/csrc/distributed/rpc/rpc_with_autograd.h>
+#include <c10/util/C++17.h>
+#include <torch/csrc/distributed/rpc/utils.h>
+#include <torch/csrc/utils/byte_order.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+constexpr int kAutogradMessageSize = 17;
+
+AutogradMetadata::AutogradMetadata(
+    int64_t autogradContextId_,
+    int64_t autogradMessageId_)
+    : autogradContextId(autogradContextId_),
+      autogradMessageId(autogradMessageId_) {}
+
+RpcWithAutograd::RpcWithAutograd(
+    MessageType messageType,
+    const AutogradMetadata& autogradMetadata,
+    std::unique_ptr<RpcCommandBase> wrappedRpc)
+    : messageType_(messageType), autogradMetadata_(autogradMetadata) {
+  TORCH_INTERNAL_ASSERT(wrappedRpc != nullptr, "wrappedRpc cannot be null!");
+  TORCH_INTERNAL_ASSERT(
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_REQ ||
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_RESP);
+  wrappedMessage_ = std::move(*wrappedRpc).toMessage();
+  tensors_ = wrappedMessage_.tensors();
+  wrappedMessageType_ = wrappedMessage_.type();
+}
+
+RpcWithAutograd::RpcWithAutograd(
+    MessageType messageType,
+    const AutogradMetadata& autogradMetadata,
+    std::unique_ptr<RpcCommandBase> wrappedRpc,
+    MessageType wrappedMessageType,
+    std::vector<torch::Tensor> tensors)
+    : messageType_(messageType),
+      autogradMetadata_(autogradMetadata),
+      wrappedRpc_(std::move(wrappedRpc)),
+      wrappedMessageType_(wrappedMessageType),
+      tensors_(std::move(tensors)) {
+  TORCH_INTERNAL_ASSERT(wrappedRpc_ != nullptr, "wrappedRpc cannot be null!");
+  TORCH_INTERNAL_ASSERT(
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_REQ ||
+      messageType_ == MessageType::MESSAGE_WITH_AUTOGRAD_RESP);
+}
+
+Message RpcWithAutograd::toMessage() && {
+  auto messageId = wrappedMessage_.id();
+  auto messageType = wrappedMessage_.type();
+
+  auto payload = std::move(wrappedMessage_).movePayload();
+  TORCH_INTERNAL_ASSERT(!payload.empty());
+
+  // We append the message type (1 byte), autograd context id(8 bytes) and
+  // autograd message id(8 bytes) to the original message in network byte order
+  // (big endian).
+  size_t writableIndex = payload.size();
+
+  // Need 17 additional bytes.
+  payload.resize(payload.size() + kAutogradMessageSize);
+
+  // Add message type.
+  payload[writableIndex++] = messageType;
+
+  // Add autograd ids.
+  torch::utils::THP_encodeInt64Buffer(
```

Comment:
Do we have a plan/solution to avoid directly manipulating bytes? At least I think we can use jit pickler to pack those fields into one payload and append it in one shot. This could also help us avoid exposing `THP_encodeInt64Buffer`.